### DISCUSSION TITLE
How to query to get the contact list which were deleted?

### DISCUSSION THREAD

**Discussion by heroickunal** [vestrel00/contacts-android#207] [Q&A] [answered]
How to query to get the contact list which were deleted?

**Comment by vestrel00** [comment, marked as answer]
Hey @heroickunal, great question! It's a tricky question to answer because there are so many different things that can be said about it. There are two ways of answering this question. Read the short answer for the TLDR. Read the long answer if you want a more detailed explanation. 

I will answer this as thoroughly as I can because it is actually a very good question whose answer is not very straightforward.

## Short answer

You cannot get contacts that have been deleted just by using the `ContactsContract` alone. Since this library is only a wrapper around `ContactsContract`, it does not allow you to get deleted contacts because they are actually deleted from the local database tables. 

In order for you to be able to get "deleted contacts", you would have to interface with accounts that support "trash". This requires network access and authentication, both of which are not in scope of this library.

## Long answer

First, we'll take a look at what other apps do with regards to deleting contacts. Second, I will explain what happens when deleting contacts using `ContactsContract`. Third, I'll talk a bit about the deleted contacts table.

#### I. What other apps do with regards to deleting contacts

Let's take a look at what the [Google Contacts](https://play.google.com/store/apps/details?id=com.google.android.contacts&hl=en_US&gl=US) app, Samsung Contacts, and the default Android Open Source Project (AOSP) Contacts app does with deleted contacts...

| **Google Contacts synced**                                                                                                     | **Google Contacts device-only**                                                                                                      | **Samsung Contacts device-only**                                                                                             | **AOSP Contacts device-only**                                                                                 |
|--------------------------------------------------------------------------------------------------------------------------------|--------------------------------------------------------------------------------------------------------------------------------------|------------------------------------------------------------------------------------------------------------------------------|---------------------------------------------------------------------------------------------------------------|
| ![google-contacts-trash](https://user-images.githubusercontent.com/4332753/166724452-bf700d11-3069-4845-9367-470f69ab2294.gif) | ![google-contacts-trash-local](https://user-images.githubusercontent.com/4332753/166725962-c7493193-29ed-4072-8483-830a88fe1e7e.gif) | ![samsung-trash-local](https://user-images.githubusercontent.com/4332753/166740750-2e693b9e-3f4b-4808-8144-38b81847dfd7.gif) | ![aosp](https://user-images.githubusercontent.com/4332753/166730915-27ce846f-7c5c-4137-a587-22c2c07b9beb.gif) |

When deleting a **synced contact** (associated with an Account; i.e. hoalwking@gmail.com), the Google Contacts app shows you a prompt to _"Move to Trash? This contact will be removed from all your synced devices"_.

There is one keyword here; "synced". You can read more about the documentation I wrote about "syncing contact data" in this page; https://vestrel00.github.io/contacts-android/entities/sync-contact-data/

Deleting the synced contact does remove it from the **local database**. However, since the synced contact is associated with a Google account, Google still has a record of the contact in their **remote database**. This is how the Google Contacts app is able to show deleted contacts- they download it over the network. It is also the same mechanism used for "back-up and restore" functionality. Turning off network connection by using airplane mode shows you that contacts in the Trash cannot be shown unless you have internet connection.

When deleting a local, device-only contact, Google Contacts shows you a prompt to _"Delete contact? This contact will be permanently deleted from your device"_. Notice that the messaging is different in this case. The contact will actually be deleted. It will not be accessible in the trash because it is not associated with a Google account. It is stored only locally in the device. If you look at the `ContactsContract.RawContacts` table before deleting the (raw) contact, you will see that it has null values for the account name and type columns.

Samsung phones replaces the AOSP Contacts app with their own proprietary Contacts app. You can enable the trash functionality there. Unlike the Google Contacts app, the Samsung Contacts app supports deleting device-only local contacts and having it live in the trash.  If you look at the `ContactsContract.RawContacts` table before deleting the (raw) contact, you will see that it uses `vnd.sec.contact.phone` as the account. This is a local account on the phone that is managed by the Samsung Contacts app. The deleted contact is actually deleted from the `ContactsContract` tables. Samsung is just keeping a copy in their own local database tables, outside the scope of `ContactsContract`. 

The default AOSP Contacts app that comes with vanilla Android phones does not support trash functionality at all for synced and local contacts due to the limitations of the `ContactsContract`. This project of mine also faces the same limitation.

#### II. How deleting contacts work when using this library or the `ContactsContract` directly

I've written consumer-facing documentation about contacts deletion here; https://vestrel00.github.io/contacts-android/basics/delete-contacts/#contacts-are-deleted-automatically-when-all-constituent-rawcontacts-are-deleted

I'll explain the internals a bit. A Contact may consist of one or more RawContacts. Deleting a RawContact does not necessarily mean that the parent Contact will be deleted. Only when all of the RawContacts have been deleted (or marked for deletion) will the Contact also be deleted. So, I have been purposefully vague in the previous section of this post when I refer to "contact" (lowercase c). 

RawContacts are either associated with an Account or they are not. RawContacts associated with an Account (e.g. a Google Account) are stored and synced with the **local and remote databases**. RawContacts with no Account (device-only, local) are only stored in the **local database**.

- When deleting RawContacts with an Account, they are marked for deletion until the next sync occurs when the corresponding SyncAdapter for that Account will perform the actual deletion in the local database. This requires a network connection. 
    - The query APIs provided by this library (e.g. `Query` and `BroadQuery`) does NOT return RawContacts that are marked for deletion even though they are still in the RawContacts table. Even though it is possible to get RawContacts pending deletion, it will not be a valid entity because the parent Contact may have already been deleted. This is the same behavior as all other contacts apps; Google Contacts, Samsung Contacts,and  AOSP Contacts.
- When deleting local, device-only RawContacts, they are deleted immediately by the Contacts Provider without a sync. No network connection is required. 

#### III. The not-so-useful `DeletedContacts` table

The [`ContactsContract.DeletedContacts`](https://developer.android.com/reference/android/provider/ContactsContract.DeletedContacts) table has been added since API 18. The table has the following columns;

```java
// Constants for the deleted contact table. This table holds a log of deleted contacts.
DeletedContactsColumns  {
    // A reference to the {@link ContactsContract.Contacts#_ID} that was deleted.
    public static final String CONTACT_ID = "contact_id";
    
    // Time (milliseconds since epoch) that the contact was deleted.
    public static final String CONTACT_DELETED_TIMESTAMP = "contact_deleted_timestamp";
}
```

This table does not contain any contact data. We only get the Contacts table row id and when it was deleted. There isn't really anything we can do with this info. It's just a log of the Contacts table row deletions that have occurred.

**Reply by vestrel00** [reply]
@heroickunal, if the above answered your question, please mark it as the answer 😁

**Reply by vestrel00** [reply]
It's been a month. I think I got the answer so I'll mark it as such 😁
